# Genetech Lab Course

## For Whom?

This repo contains all the material for the computer labs associated with the
course CB2040.

You will find the all the essential information regarding these labs within this
repo, such as guidelines regarding the execution of the labs, contact
information, deadlines and brief descriptions of the labs with links to material
that might be of help to you.

Programming and this type of analysis is best learnt by doing; hence we
encourage and strongly recommend that you work alone. But you may work in pairs
if that is what you prefer, however groups of more than two people are **not**
accepted. If you do work in pairs, please state who you are working with in the
report and comment section of your hand in, also make sure BOTH of you hand in a
copy of the report (identical).

## :dna: Introduction :dna:

**Main Responsible TA:** Sami Saarenpää<br>

This lecture is an introduction to the hands on part of the course and getting everyone up to speed with working in Posit (RStudio) Cloud.

## :dna: Lab 1 | Basic R programming and markdown :dna:

**Main Responsible TA:** Solène Frapard<br>
**Lab Status:** Ready :+1: <br>
**Working file:** [labs/ex1/main.Rmd](labs/ex1/main.Rmd) <br>

Here you will familiarize yourself with some of the basic syntax and
documentation in R. This lab will lay the foundation for the two next labs,
where the focus will be more on analyzing data with the help of certain
bioinformatic packages. Thus it is imperative that you put effort into learning
how to orient yourself in R.

### Preparatory work
Before Lab 1 starts make sure that you:

Have access to Posit (RStudio) cloud!

### Hand-in

In addition to changing the author information (see Guidelines section below),
also change the variable `GRADE_MODE` from `FALSE` to `TRUE` before handing in
the lab, if your lab knitts without any complications, you have solved all the
exercises correctly. Note that this is only necessary for lab 1.

### Resources

- Rmarkdown cheat sheet: [LINK](https://raw.githubusercontent.com/rstudio/cheatsheets/master/rmarkdown-2.0.pdf)
- Learn X in Y minutes (R): [LINK](https://learnxinyminutes.com/docs/r/)
- Swirl, Learn R in R: [LINK](https://swirlstats.com/)


## :dna: Lab 2 | Data visualization and Bulk RNA-seq analysis :dna:
**Main Responsible TA:** Xiya Song<br>
**Lab Status:**  Ready :+1: <br> 
**Working file:** [labs/ex2/main.Rmd](labs/ex2/main.Rmd) <br>

RNA sequencing (RNA-seq or bulk RNA-seq) has been a golden standard in life sciences for more than a decade. 
It uses next-generation sequencing (NGS) to quantify the RNA molecules in a biological sample and it is able to find differentially expressed genes at a sample level.
It is commonly used in case-control studies, eg. to study an effect of a new drug to a tissue.
For a long period of time RNA-seq has been the only available high-throughput method to study gene expression and it has had a huge impact on molecular biology enabling researches to study genome-wide gene expression patterns in biological samples.

This lab consists of two parts. 
First you will be introduced to ggplot2, a widely-used R package for visualisation. 
This will be used in this lab, but also later on in the course and even later during your studies. 
The second part will concentrate on RNA-seq.
You will be introduced to a RNA-seq analysis pipeline including pre-processing, quality control,  normalisation and differential gene expression using DESeq2 to find differentially expressed genes.

### Preparatory work
For the best experience we recommended that you read up a bit on the following
concept, still it will be briefly explained in the lab:

* DESeq2, DE-analysis tool presented [here](https://www.youtube.com/watch?v=UFB993xufUU&t=1s)) and [here](https://www.youtube.com/watch?v=Gi0JdrxRq5s))


## :dna: Lab 3 | Single Cell RNA-seq analysis :dna:
**Main Responsible TA:** Marco Vicari<br>
**Lab Status:** Ready :+1: <br>
**Working file:** [labs/ex3/main.Rmd](labs/ex3/main.Rmd) <br>

Single-cell RNA sequencing (or scRNA-seq) has become one of the most valuable tools 
in genomics to to answer questions about molecular processes at the cellular level. 
For example, scRNA-seq has been used to define new celltypes and cell states, study 
cell to cell interactions, track differentiation processes over time, study responses 
to drugs and much more. scRNA-seq methods have also become important tools in ongoing 
research efforts to create atlases of organs in the human body in efforts such as the 
Human Cell Atlas (HCA). 

In this lab we will explore the popular Seurat R package for single-cell genomics data 
analysis. You will be introduced to some of the most basic steps in a scRNA-seq analysis
workflow including quality control, normalization, batch correction, dimensionality 
reduction, unsupervised clustering and differential expression analysis.

## :dna: Lab 4 | Spatial Transcriptomics analysis :dna:

**Main Responsible TA:** Nayanika Bhalla<br>
**Lab Status:** Ready :+1: <br>
**Working file:** [labs/ex4/main.Rmd](labs/ex4/main.Rmd) <br>

In this lab you will work with Visium data, the spatial gene expression of human
breast cancer data! While similar to single cell data in some ways, there are
also some important differences. One thing that's extremely attractive with
spatial transcriptomics data is that the very design of the method allows us to
visualize the gene expression in the physical 2D plane - which is one of the
exercises in this lab. You will also apply some of the concepts introduced in
the previous labs, in order to make sense of your spatial data.

### Preparatory work
For the best experience we recommended that you read up a bit on the following
concepts, still these concepts will be briefly explained in the lab:

* Spatial Transcriptomics (The method presented [here](https://science.sciencemag.org/content/353/6294/78))
* Dimensionality Reduction (Focus on PCA) 
* Clustering
* Matrix Factorization
* Single cell and spatial transcriptomics data integration (The method we are using is presented [here](https://www.nature.com/articles/s42003-020-01247-y))
